### PR TITLE
@AutoDelegate: Support Interface Statics and Generics

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- Add support for static methods in interfaces
- Add basic support for generic types
- These features are generally useful; the latter is needed for the coordination store in transactions2

**Implementation Description (bullets)**:
- Filter static methods out when trying to pick which methods we want to generate a copy of
- Add correct type parameter elements to the javapoet builder when generating an AutoDelegate class

**Testing (What was existing testing like?  What have you done to improve it?)**:
There were some pretty good tests, but none for these specific behaviours. I've added new ones for these.

**Concerns (what feedback would you like?)**:
- Some more complex interactions with generics, particularly around subclassing don't work correctly. For example, if I have `interface One<A, B>` and `interface Two<C> extends One<Integer, C>`, methods from `One` generated in `AutoDelegate_Two` reference `A` and `B` which clearly don't exist. Even worse, if I have `interface One<A, B>` and `interface Two<B, A> extends One<A, B>`, the parameters for the `One` methods work opposite of the way they should.

**Where should we start reviewing?**: `GenericsTester` for an overview of what's broken, then `AutoDelegateProcessor`

**Priority (whenever / two weeks / yesterday)**: this week

### CODE REVIEW

File: atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateGenericTests.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import org.junit.Test;
+
+public class AutoDelegateGenericTests {
+    @Test
+    public void generatedInterfaceHasInterfaceMethods() {
+        Set<String> generatedMethods = TestingUtils.extractMethods(AutoDelegate_GenericsTester.class);
+        Set<String> originalMethods = extractNonStaticMethods(GenericsTester.class);
+
+        assertThat(generatedMethods, hasItems(originalMethods.toArray(new String[0])));
```
Comment (felixdesouza) [inline atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateGenericTests.java]: please use assertJ for this:

```java
assertThat(generatedMethods.contains(originalMethods);
```

File: atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateGenericTests.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Test
+    public void generatedInterfaceHasDelegateMethod() {
+        Set<String> generatedMethods = TestingUtils.extractMethods(AutoDelegate_GenericsTester.class);
+        Set<String> originalMethods = extractNonStaticMethods(GenericsTester.class);
+
+        generatedMethods.removeAll(originalMethods);
```
Comment (felixdesouza) [inline atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateGenericTests.java]: similarly here:
```java
assertThat(Sets.difference(generatedMethods, originalMethods))
    .containsOnly("delegate");
```

File: atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateInterfaceTests.java [hunk 3]
```diff
@@ -96,4 +107,10 @@ public void generatedInterfaceCallsMethodOnDelegate() {
         instanceOfInterface.methodWithReturnType();
         verify(mockImpl, times(1)).methodWithReturnType();
     }
+
+    private Set<String> extractNonStaticMethods(Class klass) {
```
Comment (felixdesouza) [inline atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateInterfaceTests.java]: nit: `static`

File: atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateGenericTests.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        assertThat(generatedMethods.size(), is(1));
+        assertThat(generatedMethods.iterator().next(), containsString("delegate"));
+    }
+
+    private Set<String> extractNonStaticMethods(Class klass) {
```
Comment (felixdesouza) [inline atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateGenericTests.java]: nit: `static`

File: atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateInterfaceTests.java [hunk 5]
```diff
@@ -63,26 +65,35 @@ public void packagePrivateInterfacesGeneratePackagePrivateInterfaces() {
...
         assertThat(generatedMethods.size(), is(1));
         assertThat(generatedMethods.iterator().next(), containsString("delegate"));
     }
 
+    @Test
+    public void generatedInterfaceDoesNotHaveStaticMethods() {
+        Set<String> generatedMethods = TestingUtils.extractMethods(AutoDelegate_TestInterface.class);
+        Set<String> originalStaticMethods = TestingUtils.extractMethodsSatisfyingPredicate(TestInterface.class,
+                method -> Modifier.isStatic(method.getModifiers()));
+
```
Comment (felixdesouza) [inline atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateInterfaceTests.java]: `assertThat(Sets.intersection(generatedMethods, originalMethods)).isEmpty()`?

File: atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateGenericTests.java (hunk 4 above)
Comment (felixdesouza) [reply atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateGenericTests.java]: perhaps move this to `TestingUtils`?

File: atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateGenericTests.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateGenericTests.java]: Ah, ok. I copied this from the other classes (where I'd prefer to stick with old-style as I'm not feeling up to changing all the old tests to use AssertJ), but no reason to hang on to this for new classes.

Comment (jeremyk-91) [reply atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateGenericTests.java]: nit: we want `containsAll`, `contains` is varargs on elements

File: atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateGenericTests.java (hunk 4 above)
Comment (jeremyk-91) [reply atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateGenericTests.java]: Yep, gets rid of the duplication as well

File: atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateInterfaceTests.java (hunk 5 above)
Comment (jeremyk-91) [reply atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateInterfaceTests.java]: Yep, that works. I'll stick with old-style here as having both junit and assertj styles in the same file gets annoying for imports.

File: atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateInterfaceTests.java (hunk 3 above)
Comment (jeremyk-91) [reply atlasdb-processors-tests/src/test/java/com/palantir/processors/AutoDelegateInterfaceTests.java]: (This method was moved to `TestingUtils`, and is static there.)

**Pull request merged**